kernel_module/legion-laptop.c

```diff
@@ -2842,84 +2875,93 @@ struct WMIFanTableRead {
-		//fancurve->points[10].speed1 = fantable->FSSA;
+		fancurve->points[0].speed1 = fantable.FSS0;
+		fancurve->points[1].speed1 = fantable.FSS1;
+		fancurve->points[2].speed1 = fantable.FSS2;
+		fancurve->points[3].speed1 = fantable.FSS3;
+		fancurve->points[4].speed1 = fantable.FSS4;
+		fancurve->points[5].speed1 = fantable.FSS5;
+		fancurve->points[6].speed1 = fantable.FSS6;
+		fancurve->points[7].speed1 = fantable.FSS7;
+		fancurve->points[8].speed1 = fantable.FSS8;
+		fancurve->points[9].speed1 = fantable.FSS9;
+		//fancurve->points[10].speed1 = fantable.FSSA;
 	}
 	return err;
 }
 
 static ssize_t wmi_write_fancurve_custom(const struct model_config *model,
 					 const struct fancurve *fancurve)
 {
-	u8 buffer[0x20];
+	struct WMIFanTable fan;
+	fan.FSTL = 10;
+	fan.FSS0 = fancurve->points[0].speed1;
+	fan.FSS1 = fancurve->points[1].speed1;
+	fan.FSS2 = fancurve->points[2].speed1;
+	fan.FSS3 = fancurve->points[3].speed1;
+	fan.FSS4 = fancurve->points[4].speed1;
+	fan.FSS5 = fancurve->points[5].speed1;
+	fan.FSS6 = fancurve->points[6].speed1;
+	fan.FSS7 = fancurve->points[7].speed1;
+	fan.FSS8 = fancurve->points[8].speed1;
+	fan.FSS9 = fancurve->points[9].speed1;
+	fan.FSSNULL = 0;
+	// TODO: These should be variable and may be set on later models
+	// on Legion Go, they are hardcoded
+	fan.FTLE = 10;
+	fan.FST0 = 10;
+	fan.FST1 = 20;
+	fan.FST2 = 30;
+	fan.FST3 = 40;
+	fan.FST4 = 50;
+	fan.FST5 = 60;
+	fan.FST6 = 70;
+	fan.FST7 = 80;
+	fan.FST8 = 90;
+	fan.FST9 = 100;
+	fan.FSTNULL = 0;
 	int err;
 
-	// The buffer is read like this in ACPI firmware
-	//
-	// CreateByteField (Arg2, Zero, FSTM)
-	// CreateByteField (Arg2, One, FSID)
-	// CreateDWordField (Arg2, 0x02, FSTL)
-	// CreateByteField (Arg2, 0x06, FSS0)
-	// CreateByteField (Arg2, 0x08, FSS1)
-	// CreateByteField (Arg2, 0x0A, FSS2)
-	// CreateByteField (Arg2, 0x0C, FSS3)
-	// CreateByteField (Arg2, 0x0E, FSS4)
-	// CreateByteField (Arg2, 0x10, FSS5)
-	// CreateByteField (Arg2, 0x12, FSS6)
-	// CreateByteField (Arg2, 0x14, FSS7)
-	// CreateByteField (Arg2, 0x16, FSS8)
-	// CreateByteField (Arg2, 0x18, FSS9)
-
-	memset(buffer, 0, sizeof(buffer));
-	buffer[0x06] = fancurve->points[0].speed1;
-	buffer[0x08] = fancurve->points[1].speed1;
-	buffer[0x0A] = fancurve->points[2].speed1;
-	buffer[0x0C] = fancurve->points[3].speed1;
-	buffer[0x0E] = fancurve->points[4].speed1;
-	buffer[0x10] = fancurve->points[5].speed1;
-	buffer[0x12] = fancurve->points[6].speed1;
-	buffer[0x14] = fancurve->points[7].speed1;
-	buffer[0x16] = fancurve->points[8].speed1;
-	buffer[0x18] = fancurve->points[9].speed1;
-
 	print_hex_dump(KERN_INFO, "legion_laptop fan table wmi write buffer",
-		       DUMP_PREFIX_ADDRESS, 16, 1, buffer, sizeof(buffer),
+		       DUMP_PREFIX_ADDRESS, 16, 1, &fan, sizeof(fan),
 		       true);
 	err = wmi_exec_arg(WMI_GUID_LENOVO_FAN_METHOD, 0,
-			   WMI_METHOD_ID_FAN_SET_TABLE, buffer, sizeof(buffer));
+			   WMI_METHOD_ID_FAN_SET_TABLE, &fan, sizeof(fan));
```

Comment:
In the old code we write a buffer of size 0x20. Now we write an object of size 0x34. Will this break stuff in old models?